### PR TITLE
print-sys-tz: Add print-system-timezone option and automatic timezone check

### PR DESCRIPTION
**Pull request by St0rmingBr4in**
This adds the `ask-before-change` option which asks the user if he wants to apply the timezone change or not.
This is particularly useful when combined with the `09-timezone` script which automatically launches tzupdate upon network connection.
The user is therefore prompted to change the system timezone whenever the system connects to a network in a different timezone.

Signed-off-by: Julien DOCHE <julien.doche@gmail.com>

### CODE REVIEW

**Comment by cdown** [comment]
Sorry, I'm not going to add a dependency on gobject, tzupdate is supposed to be minimal.

**Comment by cdown** [comment]
To be clear: if you're interested on a version of this that doesn't require being dependent on gobject, and doesn't opt all tzupdate users into the change, then that's something more palatable :-)

**Comment by St0rmingBr4in** [comment]
I am looking into having something work with only `dbus-python` as dependency, but that means I have to write some sort of minimal mainloop to receive the dbus ActionInvoked signal.

**Comment by cdown** [comment]
To be clear, dbus as a dependency would be way too much. This would need to happen externally to tzupdate.

**Comment by St0rmingBr4in** [comment]
Do you mean having everything inside the script that triggers upon network connection ? It could use notify-send and dbus-monitor ?

**Comment by cdown** [comment]
A few things:

1. The 09-timezone script itself should do the heavy lifting, and should go into a contrib/ dir to make clear it's not officially supported (it won't be, since no maintainers run it).
2. tzupdate can have features to help here, but they should not complicate the existing code. An `--ask` mode is fine.
3. 09-timezone should not be installed by default on setup. It's ok if it's done as a setup extra per PEP 508.

After that, I think this would be mergeable :-)

File: tzupdate.py [hunk 1]
```diff
@@ -178,6 +178,10 @@ def link_localtime(timezone, zoneinfo_path, localtime_path):
     return zoneinfo_tz_path
 
 
+def get_sys_timezone(zoneinfo_abspath, localtime_abspath):
+    return localtime_abspath.replace(zoneinfo_abspath + os.path.sep, "", 1)
```
Comment (cdown) [inline tzupdate.py]: `os.path.commonprefix` avoids any potential shenanigans here

File: tests/e2e_tests.py [hunk 2]
```diff
@@ -33,6 +33,17 @@ def test_print_only_no_link(link_localtime_mock, deb_tz_mock):
     assert_false(deb_tz_mock.called)
 
 
+@httpretty.activate
+@mock.patch("tzupdate.write_debian_timezone")
+@mock.patch("tzupdate.link_localtime")
+def test_print_sys_tz_no_link(link_localtime_mock, deb_tz_mock):
+    setup_basic_api_response()
```
Comment (cdown) [inline tests/e2e_tests.py]: Huh? you don't call any http methods here...

In general, this increases coverage but doesn't actually test anything. Please do check `get_sys_timezone` with a unit test :-)

File: README.rst [hunk 3]
```diff
@@ -33,6 +33,12 @@ You can also see what tzupdate would do without actually doing it by passing
 ``-p``, and specify an alternative IP address by using ``-a``. This is not an
 exhaustive list of options, see ``tzupdate --help`` for that.
 
+The ``contrib/09-timezone`` script can be place in the
```
Comment (cdown) [inline README.rst]: Let's leave this out of the README until there are known users, although it's fine to add it into a `contrib/README.rst`.

File: contrib/09-timezone [hunk 4]
```diff
@@ -0,0 +1,44 @@
+#!/bin/sh
```
Comment (cdown) [inline contrib/09-timezone]: shellcheck is legitimately fairly unhappy with this file. Please make sure it's shellcheck clean :-)

File: tests/e2e_tests.py (hunk 2 above)
Comment (cdown) [inline tests/e2e_tests.py]: The comment was marked as resolved, but again, why are we doing `setup_basic_api_response`? `--print-system-timezone` shouldn't need to reach out to get an API response

File: tests/unit_tests.py [hunk 5]
```diff
@@ -34,6 +35,13 @@ def test_get_timezone_for_ip(ip, service):
     fake_queue.put.assert_called_once_with(FAKE_TIMEZONE)
 
 
+def test_get_sys_timezone():
```
Comment (cdown) [inline tests/unit_tests.py]: :+1:

File: tzupdate.py [hunk 6]
```diff
@@ -178,6 +178,12 @@ def link_localtime(timezone, zoneinfo_path, localtime_path):
     return zoneinfo_tz_path
 
 
+def get_sys_timezone(zoneinfo_abspath, localtime_abspath):
+    return localtime_abspath.replace(
+        os.path.commonprefix([zoneinfo_abspath, localtime_abspath]) + os.path.sep, "", 1
```
Comment (cdown) [inline tzupdate.py]: :+1:

File: tests/e2e_tests.py (hunk 2 above)
Comment (St0rmingBr4in) [reply tests/e2e_tests.py]: Yeah, it was an oversight on my part, sorry

**Comment by cdown** [comment]
19283ac36760d1725eacea0e99f723ff952a239a takes care of the remaining stuff. Thanks for your contribution!

**Pull request merged**